### PR TITLE
Lpfg 522 manage canc policy

### PR DESCRIPTION
**Pull request by mick-jones**
*** Do not merge ***

- added Controller/Factory/Validator for:
  - cancellationPolicy
  - termsAndConditions
- removed t&c/cancellation policy buttons from add-learning-provider page and added them to learning-provider-overview
- added templates for:
  - add-terms-and-conditions
  - add-cancellation-policy
  - learning-provider-overview
- added components for listing:
  - cancellation policies 
  - terms and conditions 
- when setting a learning provider (name) page now redirects to the overview page
- added new right hand menu component for delete buttons (functionality to be added)
- refactored heading on some page to sit inside a full width column
- added textarea with error component
-fixed back link on add-terms-and-conditions page

### CODE REVIEW

File: src/controllers/LearningProvider/learningProviderController.ts [hunk 1]
```diff
@@ -64,9 +70,19 @@ export class LearningProviderController {
...
+	}
 
-			response.redirect('/content-management/learning-providers')
+	private async getLearningProviderAndRenderTemplate(request: Request, response: Response, view: string) {
+		const learningProviderId: string = request.params.learningProviderId
+		const learningProvider = await this.learningCatalogue.getLearningProvider(learningProviderId)
+		if (learningProvider) {
+			response.render(view, {learningProvider})
+		} else {
+			response.sendStatus(404)
```
Comment (superafroman) [inline src/controllers/LearningProvider/learningProviderController.ts]: Duplicated from CancellationPolicyController. Should leverage `app.param` from express and remove duplication.

File: src/controllers/LearningProvider/termsAndConditionsController.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		return async (request: Request, response: Response) => {
+			const data = {
+				...request.body,
+			}
+
+			const termsAndConditions = this.termsAndConditionsFactory.create(data)
+
+			const errors = await this.termsAndConditionsValidator.check(request.body, ['title'])
+			if (errors.size) {
+				return response.render('page/add-terms-and-conditions', {
```
Comment (superafroman) [inline src/controllers/LearningProvider/termsAndConditionsController.ts]: Check out post/redirect/get pattern - https://en.wikipedia.org/wiki/Post/Redirect/Get

File: src/controllers/LearningProvider/termsAndConditionsController.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		}
+	}
+
+	private async getLearningProviderAndRenderTemplate(request: Request, response: Response, view: string) {
+		const learningProviderId: string = request.params.learningProviderId
+		const learningProvider = await this.learningCatalogue.getLearningProvider(learningProviderId)
+		if (learningProvider) {
+			response.render(view, {learningProvider})
+		} else {
+			response.sendStatus(404)
```
Comment (superafroman) [inline src/controllers/LearningProvider/termsAndConditionsController.ts]: As above, duplicated.

File: src/learning-catalogue/model/course.ts [hunk 4]
```diff
@@ -36,14 +36,12 @@ export class Course {
...
-		return costArray.length
-			? costArray.reduce((p, c) => (p || 0) + (c || 0), 0)
-			: null
+		return costArray.length ? costArray.reduce((p, c) => (p || 0) + (c || 0), 0) : null
 	}
 
 	getType() {
 		if (!this.modules.length) {
-			return null
+			return 'Course'
```
Comment (superafroman) [inline src/learning-catalogue/model/course.ts]: Consistency - should be lowercase

File: src/learning-catalogue/model/factory/cancellationPolicyFactory.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import {CancellationPolicy} from '../cancellationPolicy'
+
+export class CancellationPolicyFactory {
+	constructor() {
+		this.create = this.create.bind(this)
```
Comment (superafroman) [inline src/learning-catalogue/model/factory/cancellationPolicyFactory.ts]: Is this needed?

File: src/learning-catalogue/model/factory/learningProviderFactory.ts [hunk 6]
```diff
@@ -10,6 +10,8 @@ export class LearningProviderFactory {
 
 		learningProvider.id = data.id
 		learningProvider.name = data.name
+		learningProvider.cancellationPolicies = data.cancellationPolicies
+		learningProvider.termsAndConditions = data.termsAndConditions
```
Comment (superafroman) [inline src/learning-catalogue/model/factory/learningProviderFactory.ts]: Do these need to be converted to the relevant objects?

File: src/learning-catalogue/service/termsAndConditionsService.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import {RestService} from './restService'
+import {TermsAndConditionsFactory} from '../model/factory/termsAndConditionsFactory'
+import {TermsAndConditions} from '../model/termsAndConditions'
+
+export class TermsAndConditionsService {
```
Comment (superafroman) [inline src/learning-catalogue/service/termsAndConditionsService.ts]: These services could all be generic - create an instance passing a factory and a resource path. Need to consider variable replacement in the paths but there's a lot of duplication going on that could be simplified (maybe?).

File: src/learning-catalogue/validator/termsAndConditionsValidator.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * if (errors.size()) {
+ *   for (message in errors.fields.title) {
+ *     console.log(message)
+ *   }
+ * }
+
+ */
+
+export class TermsAndConditionsValidator {
```
Comment (superafroman) [inline src/learning-catalogue/validator/termsAndConditionsValidator.ts]: Again, these classes are all the same - why duplicate?

File: views/component/policyList/template.njk [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<ul class="list">
+    <li class="provider-headers"><span class="govuk-body first-header">Title</span><span class="govuk-body">Added on</span></li>
+
+    {% for policy in learningProvider.cancellationPolicies %}
+    <li class="list-details plist-details u--border--bgrey">
+        <p class="govuk-body list-item list-name" href="/content-management/course/{{ policy.id }}">{{ policy.name }}</p>
+        <p class="govuk-body list-item list-date">13th August 2018</p>
```
Comment (superafroman) [inline views/component/policyList/template.njk]: Assume this is in a different story?

File: views/component/providerList/template.njk [hunk 10]
```diff
@@ -1,12 +1,11 @@
...
 <ul class="list">
     <li class="provider-headers"><span class="govuk-body first-header">Learning provider name</span><span class="govuk-body">Added on</span></li>
     {% for learningProvider in pageResults.results %}
 
     <li class="list-details plist-details u--border--bgrey">
-        <p class="govuk-body list-item list-name" href="/content-management/course/{{ learningProvider.id }}">{{ learningProvider.name }}</p>
-        <p class="govuk-body list-item list-date">24th Jan 2018</p>
-        <a class="govuk-link list-link " href="#">Edit</a>
+        <p class="govuk-body list-name">{{ learningProvider.name }}</p>
+        <p class="govuk-body list-item list-date">13th August 2018</p>
```
Comment (superafroman) [inline views/component/providerList/template.njk]: As above, should this date be dynamic?

File: views/component/policyList/template.njk (hunk 9 above)
Comment (m24murray) [reply views/component/policyList/template.njk]: Yes, fixing in later branch

File: views/component/providerList/template.njk (hunk 10 above)
Comment (m24murray) [reply views/component/providerList/template.njk]: Correct, but info wasnt available in Learning catalogue at the time of this branch. Will do later

File: src/learning-catalogue/model/factory/learningProviderFactory.ts (hunk 6 above)
Comment (m24murray) [reply src/learning-catalogue/model/factory/learningProviderFactory.ts]: Nope, setting directly

File: src/controllers/LearningProvider/cancellationPolicyController.ts [hunk 11]
```diff
@@ -1,37 +1,90 @@
...
+				learningProviderId,
+				cancellationPolicyId
+			)
+
+			if (cancellationPolicy) {
+				req.cancellationPolicy = cancellationPolicy
+			} else {
+				res.sendStatus(404)
+			}
+			next()
```
Comment (superafroman) [inline src/controllers/LearningProvider/cancellationPolicyController.ts]: Don't call `next()` after sending a 404.

File: src/controllers/LearningProvider/cancellationPolicyController.ts [hunk 12]
```diff
@@ -1,37 +1,90 @@
...
+			const req = ireq as ContentRequest
+
+			const learningProvider = await this.learningCatalogue.getLearningProvider(learningProviderId)
+
+			if (learningProvider) {
+				req.learningProvider = learningProvider
+			} else {
+				res.sendStatus(404)
+			}
+			next()
```
Comment (superafroman) [inline src/controllers/LearningProvider/cancellationPolicyController.ts]: As above.

File: src/controllers/LearningProvider/learningProviderController.ts [hunk 13]
```diff
@@ -1,41 +1,68 @@
...
+			const req = ireq as ContentRequest
+
+			const learningProvider = await this.learningCatalogue.getLearningProvider(learningProviderId)
+
+			if (learningProvider) {
+				req.learningProvider = learningProvider
+			} else {
+				res.sendStatus(404)
+			}
+			next()
```
Comment (superafroman) [inline src/controllers/LearningProvider/learningProviderController.ts]: As above.

File: src/controllers/LearningProvider/termsAndConditionsController.ts [hunk 14]
```diff
@@ -1,37 +1,90 @@
...
+				learningProviderId,
+				termsAndConditionsId
+			)
+
+			if (termsAndConditions) {
+				req.termsAndConditions = termsAndConditions
+			} else {
+				res.sendStatus(404)
+			}
+			next()
```
Comment (superafroman) [inline src/controllers/LearningProvider/termsAndConditionsController.ts]: As above.

File: src/controllers/LearningProvider/termsAndConditionsController.ts [hunk 15]
```diff
@@ -1,37 +1,90 @@
...
+			const req = ireq as ContentRequest
+
+			const learningProvider = await this.learningCatalogue.getLearningProvider(learningProviderId)
+
+			if (learningProvider) {
+				req.learningProvider = learningProvider
+			} else {
+				res.sendStatus(404)
+			}
+			next()
```
Comment (superafroman) [inline src/controllers/LearningProvider/termsAndConditionsController.ts]: As above.

File: src/controllers/LearningProvider/termsAndConditionsController.ts [hunk 16]
```diff
@@ -1,37 +1,90 @@
...
+			this.setTermsAndConditions()
+		)
 	}
 
 	public getTermsAndConditions() {
 		logger.debug('Getting terms and conditions')
 		return async (request: Request, response: Response) => {
-			await this.getLearningProviderAndRenderTemplate(request, response, 'page/add-terms-and-conditions')
+			const req = request as ContentRequest
+			const learningProvider = req.learningProvider
```
Comment (superafroman) [inline src/controllers/LearningProvider/termsAndConditionsController.ts]: Could avoid this here and in all of these get functions by just sticking `learningProvider` on `req.locals` in the param filter.

**Pull request merged**